arch-riscv: `vred*` fails in `assert(arch_reg.getNumPinnedWrites() == 0)`

**Describe the bug**
The `vred*` instructions of RVV creates a `vpin_micro` instructions, which can cause the following instructions to fail in the `assert(arch_reg.getNumPinnedWrites() == 0)` of `rename_map.cc` file.

**Affects version**
I'm using the `develop` branch in commit `59f093c146098ad5c72e657d57534c6a8d92ca41`.

**gem5 Modifications**
I'm using my modifications from this PR [#2657](https://github.com/gem5/gem5/pull/2657), since without them the bug reported here [#2621](https://github.com/gem5/gem5/issues/2621) will happen.

**To Reproduce**
I'm using `gem5.opt` compiled with `scons build/RISCV/gem5.opt -j20`.
In special, when running the code use `VLEN=512`.

My code was compiled with this command:

```
riscv64-unknown-linux-musl-gcc -O2 -static -Wall -Wextra -o build/app_bin main.c compress.c
```

`main.c` file:

```
#include <stdint.h>
#include <stdio.h>
#include <riscv_vector.h>

#define SIZE (512 + 248)

int verify(const uint8_t *a, const uint8_t *b, size_t len) {
    size_t vl = 0;
    vuint8m1_t vr = __riscv_vmv_v_x_u8m1(0, len);
    for (size_t i = 0; i < len; i += vl) {
        vl = __riscv_vsetvl_e8m8(len - i);
        vuint8m8_t va = __riscv_vle8_v_u8m8(&a[i], vl);
        vuint8m8_t vb = __riscv_vle8_v_u8m8(&b[i], vl);
        va = __riscv_vxor_vv_u8m8(va, vb, vl);
        vr = __riscv_vredor_vs_u8m8_u8m1(va, vr, vl);
    }
    uint8_t r = __riscv_vmv_x_s_u8m1_u8(vr);

    return r;
}

int main() {
  uint8_t a[SIZE], b[SIZE];

  for(int i = 0; i < SIZE; i++) {
    a[i] = 2*i;
    b[i] = 2*i;
  }

  printf("Equal: %d\n", verify(a, b, SIZE));

  return 0;
}
```

**Terminal Output**

```shell
Beginning simulation!
gem5.opt: src/cpu/o3/rename_map.cc:84: gem5::o3::SimpleRenameMap::RenameInfo gem5::o3::SimpleRenameMap::rename(const gem5::RegId&): Assertion `arch_reg.getNumPinnedWrites() == 0' failed.
Program aborted at tick 2310000
--- BEGIN LIBC BACKTRACE ---
gem5.opt(_ZN4gem515print_backtraceEv+0x30)[0x59345e0b8860]
gem5.opt(_ZN4gem512abortHandlerEi+0x4c)[0x59345e0e135c]
/lib/x86_64-linux-gnu/libc.so.6(+0x45330)[0x7e8fa6045330]
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x11c)[0x7e8fa609eb2c]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x1e)[0x7e8fa604527e]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xdf)[0x7e8fa60288ff]
/lib/x86_64-linux-gnu/libc.so.6(+0x2881b)[0x7e8fa602881b]
/lib/x86_64-linux-gnu/libc.so.6(+0x3b517)[0x7e8fa603b517]
gem5.opt(_ZN4gem52o315SimpleRenameMap6renameERKNS_5RegIdE+0x9f7)[0x59345d7856a7]
gem5.opt(_ZN4gem52o36Rename14renameDestRegsERKNS_14RefCountingPtrINS0_7DynInstEEEs+0x25d)[0x59345d77c66d]
gem5.opt(_ZN4gem52o36Rename11renameInstsEs+0x4ce)[0x59345d77f61e]
gem5.opt(_ZN4gem52o36Rename6renameERbs+0x74)[0x59345d7812f4]
gem5.opt(_ZN4gem52o36Rename4tickEv+0x97)[0x59345d781787]
gem5.opt(_ZN4gem52o33CPU4tickEv+0x156)[0x59345d6f53e6]
gem5.opt(_ZN4gem510EventQueue10serviceOneEv+0xb7)[0x59345e0cf327]
gem5.opt(_ZN4gem59doSimLoopEPNS_10EventQueueE+0x70)[0x59345e100610]
gem5.opt(_ZN4gem58simulateEm+0x25e)[0x59345e100dae]
gem5.opt(+0x2248d42)[0x59345e015d42]
gem5.opt(+0x112a30a)[0x59345cef730a]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(+0x1df4b8)[0x7e8fa73df4b8]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(_PyObject_Call+0x5c)[0x7e8fa738485c]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(_PyEval_EvalFrameDefault+0x8ec)[0x7e8fa7319dec]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(PyEval_EvalCode+0x20f)[0x7e8fa74a091f]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(+0x29c8b0)[0x7e8fa749c8b0]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(+0x1dfadc)[0x7e8fa73dfadc]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(PyObject_Vectorcall+0x5c)[0x7e8fa7382b2c]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(_PyEval_EvalFrameDefault+0x40ee)[0x7e8fa731d5ee]
gem5.opt(+0x222b499)[0x59345dff8499]
gem5.opt(main+0x218)[0x59345cd8c368]
/lib/x86_64-linux-gnu/libc.so.6(+0x2a1ca)[0x7e8fa602a1ca]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x8b)[0x7e8fa602a28b]
gem5.opt(_start+0x25)[0x59345cdb3645]
--- END LIBC BACKTRACE ---
For more info on how to address this issue, please visit https://www.gem5.org/documentation/general_docs/common-errors/ 

make: *** [Makefile:29: gem5] Aborted (core dumped)

```

**Expected behavior**
No `assert` fails.

**Host Operating System**
Ubuntu 24.04.2.

**Host ISA**
RISC-V.

**Compiler used**
I use riscv-gnu-toolchain in tag 2025.07.16 for linux environment (my stdlib is musl).

**Additional information**

When reading the `VectorReduceMacroConstructor` source code, I noted that the `vred*` macro-instruction always generate a `vpin_micro` instruction with the number of pinned registers equal to `LMUL`. However, in practice the number of executed instructions can be smaller than `LMUL`, for example with `VLEN = 512` the second loop iteration will only generate four micro-instructions. Then, after the four micro-instructions executing, the physical register will still have `8-4 = 4` pinned writes.

The problem is the `O3` CPU will predict that the next loop iteration will happen (miss-prediction), then creating a new speculative `vred*` instruction, producing a speculative `vpin_micro`, which will set the number of pinned writes of the architectural register to `LMUL`.

However, when renaming the destiny register of the following micro-instruction (a `vred*_micro`), the code will enter in the `else-if` of the following function from `rename_map.cc` file and it will fail in the `assert` (architectural register have 8 pinned writes and physical register have 4 pinned writes):

```
SimpleRenameMap::RenameInfo
SimpleRenameMap::rename(const RegId& arch_reg)
{
    PhysRegIdPtr renamed_reg;
    // Record the current physical register that is renamed to the
    // requested architected register.
    PhysRegIdPtr prev_reg = map[arch_reg.index()];

    if (arch_reg.is(InvalidRegClass)) {
        assert(prev_reg->is(InvalidRegClass));
        renamed_reg = prev_reg;
    } else if (prev_reg->getNumPinnedWrites() > 0) {
        // Do not rename if the register is pinned
        assert(arch_reg.getNumPinnedWrites() == 0);  // Prevent pinning the
                                                     // same register twice
        DPRINTF(Rename, "Renaming pinned reg, numPinnedWrites %d\n",
                prev_reg->getNumPinnedWrites());
        renamed_reg = prev_reg;
        renamed_reg->decrNumPinnedWrites();
    } else {
        renamed_reg = freeList->getReg();
        map[arch_reg.index()] = renamed_reg;
        renamed_reg->setNumPinnedWrites(arch_reg.getNumPinnedWrites());
        renamed_reg->setNumPinnedWritesToComplete(
            arch_reg.getNumPinnedWrites() + 1);
    }

    DPRINTF(Rename, "Renamed reg %d to physical reg %d (%d) old mapping was"
            " %d (%d)\n",
            arch_reg, renamed_reg->flatIndex(), renamed_reg->flatIndex(),
            prev_reg->flatIndex(), prev_reg->flatIndex());

    return RenameInfo(renamed_reg, prev_reg);
}
```

--- a/src/arch/riscv/isa/formats/vector_arith.isa
+++ b/src/arch/riscv/isa/formats/vector_arith.isa
@@ -44,6 +44,12 @@ let {{
     def setSrcVm():
         return "if (!this->vm)\n" + \
                "    setSrcRegIdx(_numSrcRegs++, vecRegClass[0]);"
+    def setSrcVmOrTmp0():
+        return "if (!this->vm){\n" + \
+               "(_machInst.vd == 0) ? " + \
+               "setSrcRegIdx(_numSrcRegs++,vecRegClass[VecMemInternalReg0]):"+\
+               "setSrcRegIdx(_numSrcRegs++,vecRegClass[0]);\n" + \
+               "}\n"
     def vmDeclAndReadData():
         return '''
             [[maybe_unused]] RiscvISA::vreg_t tmp_v0;
@@ -1294,7 +1300,7 @@ def format VectorReduceIntFormat(code, category, *flags) {{
     set_dest_reg_idx = setDestWrapper(dest_reg_id)
     set_src_reg_idx = setSrcWrapper(src1_reg_id)
     set_src_reg_idx += setSrcWrapper(src2_reg_id)
-    set_src_reg_idx += setSrcVm()
+    set_src_reg_idx += setSrcVmOrTmp0()
     vm_decl_rd = vmDeclAndReadData()
     set_vlenb = setVlenb()
 
@@ -1343,7 +1349,7 @@ def format VectorReduceFloatFormat(code, category, *flags) {{
     set_dest_reg_idx = setDestWrapper(dest_reg_id)
     set_src_reg_idx = setSrcWrapper(src1_reg_id)
     set_src_reg_idx += setSrcWrapper(src2_reg_id)
-    set_src_reg_idx += setSrcVm()
+    set_src_reg_idx += setSrcVmOrTmp0()
     vm_decl_rd = vmDeclAndReadData()
     set_vlenb = setVlenb()
 
@@ -1397,7 +1403,7 @@ def format VectorReduceFloatWideningFormat(code, category, *flags) {{
     set_dest_reg_idx = setDestWrapper(dest_reg_id)
     set_src_reg_idx = setSrcWrapper(src1_reg_id)
     set_src_reg_idx += setSrcWrapper(src2_reg_id)
-    set_src_reg_idx += setSrcVm()
+    set_src_reg_idx += setSrcVmOrTmp0()
     vm_decl_rd = vmDeclAndReadData()
     set_vlenb = setVlenb()
     type_def = '''
@@ -1510,7 +1516,7 @@ def format VectorReduceIntWideningFormat(code, category, *flags) {{
     set_dest_reg_idx = setDestWrapper(dest_reg_id)
     set_src_reg_idx = setSrcWrapper(src1_reg_id)
     set_src_reg_idx += setSrcWrapper(src2_reg_id)
-    set_src_reg_idx += setSrcVm()
+    set_src_reg_idx += setSrcVmOrTmp0()
     vm_decl_rd = vmDeclAndReadData()
     set_vlenb = setVlenb()
 

--- a/src/arch/riscv/isa/templates/vector_arith.isa
+++ b/src/arch/riscv/isa/templates/vector_arith.isa
@@ -1714,7 +1714,14 @@ template<typename ElemType>
         micro_vl = std::min(tmp_vl -= micro_vlmax, micro_vlmax);
     }
 
-    microop = new VPinVdMicroInst(machInst, 0, num_microops, elen, vlen);
+    int k;
+    tmp_vl = this->vl;
+    micro_vl = std::min(tmp_vl, micro_vlmax);
+    for (k = 0; k < num_microops && micro_vl > 0; ++k) {
+        micro_vl = std::min(tmp_vl -= micro_vlmax, micro_vlmax);
+    }
+
+    microop = new VPinVdMicroInst(machInst, 0, k, elen, vlen);
     microop->setFlag(IsDelayedCommit);
     this->microops.push_back(microop);
 

--- a/src/arch/riscv/isa/templates/vector_arith.isa
+++ b/src/arch/riscv/isa/templates/vector_arith.isa
@@ -1704,8 +1704,15 @@ template<typename ElemType>
         this->microops.push_back(microop);
     }
 
+    if (machInst.vd != machInst.vs1 && (machInst.vd == 0 && !machInst.vm)) {
+        microop = new VCpyVsMicroInst(machInst, 0, 0, elen, vlen);
+        microop->setFlag(IsDelayedCommit);
+        this->microops.push_back(microop);
+    }
+
     for (int i = 0; i < num_microops && micro_vl > 0; ++i) {
-        if (machInst.vd == (machInst.vs2 + i) && machInst.vd != machInst.vs1) {
+        if (machInst.vd == (machInst.vs2 + i) && machInst.vd != machInst.vs1 &&
+            !(machInst.vd == 0 && !machInst.vm)) {
             microop = new VCpyVsMicroInst(machInst, i, machInst.vs2, elen,
                                           vlen);
             microop->setFlag(IsDelayedCommit);

